### PR TITLE
Pab/infer types 2026

### PR DESCRIPTION
**Pull request by pbroadbery**
This isn't for merging just yet, just a work in progress.  It might be interesting to try some examples.

### Main changes
- Add '?' as an alias for 'figure out this type from context'.  To be used in function declarations
- Add inference mechanism for these '?' types.  
- Examples are in the test/rtypes directory

### Stuff to do
- There's going to be a lot of corner cases, so more testing
- Errors messages are profoundly awful.  Some research is needed to find a good algorithm for messages
- Remove the need for '?'

The last two of these could happen in either order but on balance the first is more useful

### Pending improvents
- "f(a: ?T1, b: ?T1): List(?T1) == ..". could be used where '?T1' refers to the same unknown type.  
- There's initial support for 'import from List' importing all variants of 'List'.  
 This needs reworking vs the new inference algorithm

### CODE REVIEW

File: aldor/aldor/src/gf_embed.c [hunk 1]
```diff
@@ -0,0 +1,262 @@
+#include "gf_embed.h"
+#include "gf_prog.h"
+#include "gf_util.h"
+#include "gf_rtime.h"
+#include "tform.h"
+#include "util.h"
+
+Foam
+gen0Embed(Foam val, AbSyn ab, TForm tf, AbEmbed embed)
```
Comment (pbroadbery) [inline aldor/aldor/src/gf_embed.c:9]: Separate commit
